docs/architecture/adr-065-custom-event-indexing.md

```diff
@@ -0,0 +1,341 @@
+  - [Status](#status)
+  - [Context](#context)
+  - [Alternative Approaches](#alternative-approaches)
+  - [Decision](#decision)
+  - [Detailed Design](#detailed-design)
+    - [EventSink](#eventsink)
+    - [Supported Sinks](#supported-sinks)
+      - [`KVEventSink`](#kveventsink)
+      - [`PSQLEventSink`](#psqleventsink)
+    - [Configuration](#configuration)
+  - [Consequences](#consequences)
+    - [Positive](#positive)
+    - [Negative](#negative)
+    - [Neutral](#neutral)
+  - [References](#references)
+
+## Changelog
+
+- April 1, 2021: Initial Draft (@alexanderbez)
+
+## Status
+
+Proposed
+
+## Context
+
+Currently, Tendermint Core supports block and transaction event indexing through
+the `tx_index.indexer` configuration. Events are captured in transactions and
+are indexed via a `TxIndexer` type. Events are captured in blocks, specifically
+from `BeginBlock` and `EndBlock` application responses, and are indexed via a
+`BlockIndexer` type. Both of these types are managed by a single `IndexerService`
+which is responsibly for consuming events and sending those events off to be
+indexed by the respective type.
+
+In addition to indexing, Tendermint Core also supports the ability to query for
+both indexed transaction and block events via Tendermint's RPC layer. The ability
+to query for these indexed events facilitates a great multitude of upstream client
+and application capabilities, e.g. block explorers, IBC relayers, and auxiliary
+data availability and indexing services.
+
+Currently, Tendermint only supports indexing via a `kv` indexer, which is supported
+by an underlying embedded key/value store database. The `kv` indexer implements
+its own indexing and query mechanisms. While the former is somewhat trivial,
+providing a rich and flexible query layer is not as trivial and has caused many
+issues and UX concerns for upstream clients and applications.
+
+The fragile nature of the proprietary `kv` query engine and the potential
+performance and scaling issues that arise when a large number of consumers are
+introduced, motivate the need for a more robust and flexible indexing and query
+solution.
+
+## Alternative Approaches
+
+With regards to alternative approaches to a more robust solution, the only serious
+contender that was considered was to transition to using [SQLite](https://www.sqlite.org/index.html).
+
+While the approach would work, it locks us into a specific query language and
+storage layer, so in some ways it's only a bit better than our current approach.
+In addition, the implementation would require the introduction of CGO into the
+Tendermint Core stack, whereas right now CGO is only introduced depending on
+the database used.
+
+## Decision
+
+We will adopt a similar approach to that of the Cosmos SDK's `KVStore` state
+listening described in [ADR-038](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-038-state-listening.md).
+
+Namely, we will perform the following:
```

Comment:
These are good points and I agree.

There was much discussion on ADR 038 about guaranteeing delivery. I am not 100% in agreement with the final decision (too many files IMO), but am fully in support of the idea. Tendermint should synchronously write to some persistent storage, which can then be processed asynchronously by a consumer. This means Tendermint doesn't slow down if the consumer blocks (or crashes), and the consumer will not lose any blocks if it restarts (or slows down).

Database replogs are probably the most advanced of this architecture (for high throughput), but every decent messaging queue can also guarantee at-least-once delivery if so desired.  We can use one of those for example, or just the same technique from ADR 038, but I agree with @iramiller below this is a very important point to specify.